modules/collect/src/main/java/com/opengamma/strata/collect/array/Array.java

```diff
@@ -0,0 +1,59 @@
+/**
+ * Copyright (C) 2016 - present by OpenGamma Inc. and the OpenGamma group of companies
+ * 
+ * Please see distribution for license.
+ */
+package com.opengamma.strata.collect.array;
+
+import java.util.stream.Stream;
+
+/**
+ * Base interface for all array types.
+ * <p>
+ * This provides an abstraction over data structures that represent an array accessed by index.
+ * <p>
+ * Implementations must be immutable and thread-safe beans.
+ * 
+ * @param <T>  the type of the element, the boxed type if primitive
+ */
+public interface Array<T> {
+
+  /**
+   * Gets the size of the array.
+   * <p>
+   * This is the total number of elements in the array.
+   * 
+   * @return the size of the array
+   */
+  public abstract int size();
+
+  /**
+   * Checks if this array is empty.
+   * 
+   * @return true if empty
+   */
+  public default boolean isEmpty() {
+    return size() == 0;
+  }
+
+  /**
+   * Gets the element at the specified index in this array.
+   * <p>
+   * The index must be valid, between zero and {@code size()}.
```

Comment:
I read this as meaning `0 <= index <= size` which is obviously wrong. "between zero and `size() - 1`"?